neo/io/nwbio.py

```diff
@@ -0,0 +1,810 @@
+                    # take single value from set
+                    annotations[annotation_name], = annotations[annotation_name]
+        if "identifier" not in annotations:
+            annotations["identifier"] = self.filename
+        if "session_description" not in annotations:
+            annotations["session_description"] = blocks[0].description or self.filename
+            # todo: concatenate descriptions of multiple blocks if different
+        if "session_start_time" not in annotations:
+            annotations["session_start_time"] = datetime.now()
+        # todo: handle subject
+        # todo: store additional Neo annotations somewhere in NWB file
+        nwbfile = NWBFile(**annotations)
+
+        assert self.nwb_file_mode in ('w',)  # possibly expand to 'a'ppend later
+        if self.nwb_file_mode == "w" and os.path.exists(self.filename):
+            os.remove(self.filename)
+        io_nwb = pynwb.NWBHDF5IO(self.filename, manager=get_manager(), mode=self.nwb_file_mode)
+
+        if sum(statistics(block)["SpikeTrain"]["count"] for block in blocks) > 0:
+            nwbfile.add_unit_column('_name', 'the name attribute of the SpikeTrain')
+            #nwbfile.add_unit_column('_description', 'the description attribute of the SpikeTrain')
+            nwbfile.add_unit_column(
+                'segment', 'the name of the Neo Segment to which the SpikeTrain belongs')
+            nwbfile.add_unit_column(
+                'block', 'the name of the Neo Block to which the SpikeTrain belongs')
+
+        if sum(statistics(block)["Epoch"]["count"] for block in blocks) > 0:
+            nwbfile.add_epoch_column('_name', 'the name attribute of the Epoch')
+            #nwbfile.add_epoch_column('_description', 'the description attribute of the Epoch')
+            nwbfile.add_epoch_column(
+                'segment', 'the name of the Neo Segment to which the Epoch belongs')
+            nwbfile.add_epoch_column('block', 'the name of the Neo Block to which the Epoch belongs')
+
+        for i, block in enumerate(blocks):
+            self.write_block(nwbfile, block)
+        io_nwb.write(nwbfile)
+        io_nwb.close()
+
+    def write_block(self, nwbfile, block, **kwargs):
+        """
+        Write a Block to the file
+            :param block: Block to be written
+        """
+        electrodes = self._write_electrodes(nwbfile, block)
+        if not block.name:
+            block.name = "block%d" % self.blocks_written
+        for i, segment in enumerate(block.segments):
+            assert segment.block is block
+            if not segment.name:
+                segment.name = "%s : segment%d" % (block.name, i)
+            self._write_segment(nwbfile, segment, electrodes)
+        self.blocks_written += 1
+
+    def _write_electrodes(self, nwbfile, block):
+        # this handles only icephys_electrode for now
+        electrodes = {}
+        devices = {}
+        for segment in block.segments:
```

Comment:
Currently in NWB it is a requirement that all times are with respect to the same session start time. I don't know how we would properly handle multiple segments.

Reply:
What about raising an Exception when writing to NWB if any segments have overlapping (t_start, t_stop)?

Reply:
Maybe I am not fully understanding the concept of a segment. Do all segments have a common time reference? If so, I don't think there is an issue with writing multiple of them to NWB.

Reply:
Segments do not necessarily have a common time reference, but they may have. If the segments' time values do not overlap, I think it is a reasonable assumption that they _do_ share a common reference.